### PR TITLE
Add summation

### PR DESCRIPTION
**Pull request by nryant**
Overview
---------
Adds following methods to `BaseMetric` to support summation:

- `__add__`
- `__radd__`

Motivation
-----------
The motivation is two-fold:

1. *multiprocessing*

Suppose we want to compute metrics for a large volume of data  --  sufficiently large that we would like to parallelize the collection of sufficient statistics. Computing the sufficient statistics using `multiprocessing` or `joblib` is straightforward; e.g., using `joblib`:

    def process_one(reference, hypothesis, uem):
       """Wrapper function for multiprocessing."""
        der = DiarizationErrorRate()
        der(reference, hypothesis, uem=uem)
        return der
    
    def main():
        # Mapping from URIs to reference annotations.
        reference_anns = ...
        
        # Mapping from URIs to hypothesis annotations.
        hypothesis_anns = ...
        
        # Mapping from URIs to scoring regions.
        uems = ...
          
        # Number of parallel processes to use.
        n_jobs = 10
        
        # Actual joblib call.
        f = delayed(process_one)
        file_ders = Parallel(n_jobs)(f(reference_anns[uri], hypothesis_anns[uri], uems[uri])
                                     for uri in reference_anns)
                                              
However, the sufficient statistics for DER computation are now spread across the file-level metrics. Combining these into a single instance reduces to:

    combined_der = sum(file_ders)
                                      
2. *aggregation*

Suppose we want to compute metrics not just overall and at a file-level, but by various logical subdivision; e.g., DIHARD III domains. This is now trivial using `pandas` dataframes. E.g. suppose `data` contains the following columns:

- `file_id`  --  file id
- `domain` --  domain file is from
- `der`  --  instance of `DiarizationErrorRate`

then:

    data.groupby('domain', as_index=False).sum()
    
will contain domain-level metrics, from which suitable reports may be generated.

### CODE REVIEW

**Comment by hbredin** [comment]
I [have](https://github.com/pyannote/pyannote-metrics/commit/70214771d6029941776085f9d9efe34092e7e7e7) [struggled](https://github.com/pyannote/pyannote-metrics/commit/4c1be0115801297533f258c94266db2d8cff17cb) in the past to make pyannote.metrics work well in a distributed way.

Your proposed solution is great! Thanks @nryant.

File: pyannote/metrics/base.py [hunk 1]
```diff
@@ -247,6 +247,21 @@ def __iter__(self):
         for uri, component in self.results_:
             yield uri, component
 
+    def __add__(self, other):
+        cls = self.__class__
+        result = cls()
```
Comment (hbredin) [inline pyannote/metrics/base.py:252]: Can we find a way to make sure `result` is initialized with the same options (e.g. `collar` and `skip_overlap` for `DiarizationErrorRate` instances) as `self` and `other`? 

This probably means adding some kind of `sklearn`-like [mechanism to clone](https://scikit-learn.org/stable/modules/generated/sklearn.base.clone.html#sklearn.base.clone) metrics.

File: pyannote/metrics/base.py [hunk 2]
```diff
@@ -247,6 +247,21 @@ def __iter__(self):
...
+        result.results_ = self.results_ + other.results_
+        for cname in self.components_:
+            result.accumulated_[cname] += self.accumulated_[cname]
+            result.accumulated_[cname] += other.accumulated_[cname]
+        return result
+
+    def __radd__(self, other):
+        if other == 0:
```
Comment (hbredin) [inline pyannote/metrics/base.py:260]: Is this how the built-in `sum` function initializes its own accumulator? 
Would be nice to add a quick comment mentionning this...

**Comment by hbredin** [comment]
Adding back a "parallel processing" entry (mentioning your joblib example, for instance) would be nice to have as well! 

https://github.com/pyannote/pyannote-metrics/commit/4c1be0115801297533f258c94266db2d8cff17cb

File: pyannote/metrics/base.py (hunk 1 above)
Comment (nryant) [reply pyannote/metrics/base.py:252]: Good point! I cribbed this PR from some monkey patching I did a couple of years ago for an internal SAD scoring tool and in that context, default parameter values were being used, so the issue didn't come up. After looking at how `sklearn` handles this, maybe we could add a similar method to ensure the resulting instance is initialized with the same arguments as the first summand. If so, I should also document that `sum([m1, m2, ...])` assumes all metrics were initialized identically (reasonable).

File: pyannote/metrics/base.py (hunk 2 above)
Comment (nryant) [reply pyannote/metrics/base.py:260]: The built-in `sum` actually has a `start` parameter that controls the initial value of the summation. As you might gather, the default value is for `start` is 0, so I just hard coded that value as an additive identity for metrics. Probably would be good to add a one or two line comment to this effect to save someone having to read up on `__radd__` and `sum`.

File: pyannote/metrics/base.py (hunk 1 above)
Comment (nryant) [reply pyannote/metrics/base.py:329]: I've made an initial attempt at a `clone` function, which also required me to implement a `get_params` method for metrics. **NOTE** that `get_params` makes the assumption that all subclasses of `BaseMetric` include `**kwargs` in their signature and pass these keyword arguments to the constructor of the superclass (or if multiple superclasses, to one of them). Should this assumption be violated, weirdness could ensue.

An alternate approach would be that used within `sklearn`, which bans use of `*args` and `**kwargs` within constructors and forces each metric to be explicit about its parameters. This would require touching more lines of the codebase, but  beyond being a bit of a chore, shouldn't be difficult to implement.

File: pyannote/metrics/base.py [hunk 3]
```diff
@@ -36,6 +37,51 @@
...
 
 
+# TODO:
+# - fit/unfit = proper terminology?
+def clone(metric):
+    """Construct a new unfit metric with the same parameters as `metric`.
+
+    Clone does a deep copy of the metric without actually copying any
+    results or accumulators. It returns a new metric with the same
+    parameters that has not been fitted on any data.
```
Comment (hbredin) [inline pyannote/metrics/base.py:47]:
```suggestion
    results or accumulators. It returns a new metric (with the same
    parameters) that has not yet been used to evaluate any data.
```

File: pyannote/metrics/base.py [hunk 4]
```diff
@@ -36,6 +37,51 @@
 from pyannote.metrics.types import Details, MetricComponents
 
 
+# TODO:
+# - fit/unfit = proper terminology?
```
Comment (hbredin) [inline pyannote/metrics/base.py:41]:
```suggestion
```

File: pyannote/metrics/base.py [hunk 5]
```diff
@@ -36,6 +37,51 @@
...
+        The metric to be cloned.
+
+    Returns
+    -------
+    metric : object
+        The deep copy of the metric,
+    """
+    if hasattr(metric, '__pyannote_clone__') and not inspect.isclass(metric):
+        # If metric implements a custom cloning method, default to that.
+        return metric.__pyannote_clone()
```
Comment (hbredin) [inline pyannote/metrics/base.py:61]:
```suggestion
        return metric.__pyannote_clone__()
```

File: pyannote/metrics/base.py [hunk 6]
```diff
@@ -36,6 +37,51 @@
 from pyannote.metrics.types import Details, MetricComponents
 
 
+# TODO:
+# - fit/unfit = proper terminology?
+def clone(metric):
```
Comment (hbredin) [inline pyannote/metrics/base.py:42]: Can we make it a method of `BaseMetric`?

File: pyannote/metrics/base.py [hunk 7]
```diff
@@ -247,6 +324,21 @@ def __iter__(self):
         for uri, component in self.results_:
             yield uri, component
 
+    def __add__(self, other):
+        cls = self.__class__
+        result = cls()
```
Comment (hbredin) [inline pyannote/metrics/base.py:329]: Once `clone` becomes a method of `BaseMetric`, we would do:

```suggestion
        result = self.clone()
```

File: tests/test_base.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+    m = M2(a=100, b=1000)
+    expected = {'a': 100, 'b': 1000}
+    actual = m.get_params()
+    assert actual == expected
+
+
+def test_clone():
+    # Tests that clone creates deep copy of "unfit" metric.
+    metric = M1(a=10)
+    metric_new = clone(metric)
```
Comment (hbredin) [inline tests/test_base.py:143]: Once `clone` is a method of `BaseMetric`:

```suggestion
    metric_new = self.clone(metric)
```

**Review by hbredin — CHANGES_REQUESTED** [review]

File: pyannote/metrics/base.py (hunk 6 above)
Comment (hbredin) [reply pyannote/metrics/base.py:40]: It needs access to `self`, doesn't it? So regular method.
Unless there is something I don't understand about the distinction between method and statcimethod...
